index.js

```diff
@@ -98,42 +106,42 @@ module.exports = (schema) => {
   });
 
   schema.pre('findOneAndRemove', function preFindOneAndRemove(next) {
```

Comment:
These functions would all need to `async` (and use `await`) as well, right? This is a little off since they use the `next()` function as well. Mongoose 5.x (I think) will see that a promise is returned and move along when the promise is resolved. But the code is written to wait for `next()` to be called. And Mongoose 4.x doesn't support promises from middleware, so there might be some inconsistent behavior.

Reply:
This should work as is, because we are just forwarding next it doesn't care about the promises.

Reply:
You've got me thinking on overdrive here. I think the issue is that if one of the lower level functions throws an exception, then we'll get an UnresolvedPromiseRejection since the promise isn't tied all the way up. `removeQuery` is an async function that will return a promise, and that promise isn't handled.

You could poke at this and test it. I suspect you might need to do 

```js
removeQuery(this, next).catch(err => next(err));
```

It's a little jank since the non-error path passes the next function down, but the error path is handled up here as a promise. This actually might be a problem in the existing code, so I'm open to you highlighting that and us creating a new task to properly handle exceptions.

Reply:
I really appreciate the thought around the error path here. I think it is important and it is something I hadn't considered.

Reply:
I'm not sure I fully understand the problem you are outlining, but I would love to learn more.

Reply:
I think I see what you are saying, does mongoose do `.catch()` on returned promises?